File: LeetCode/0227. Basic Calculator II/solution.py
"""
109 / 109 test cases passed.
Runtime: 72 ms
Memory Usage: 16.8 MB
"""
class Solution:
    def calculate(self, s: str) -> int:
        formula = s + '$'
        nums = []
        sign = '+'
        num = 0
        for c in formula:
            if c == ' ':
                continue
            if c.isdigit():
                num = num * 10 + int(c)
            else:
                if sign == '+':
                    nums.append(num)
                elif sign == '-':
                    nums.append(-num)
                elif sign == '*':
                    nums[-1] *= num
                elif sign == '/':
                    nums[-1] = int(nums[-1] / num)
                sign = c
                num = 0
        return sum(nums)

File: LeetCode/0227. Basic Calculator II/solution.cpp
/**
 * 109 / 109 test cases passed.
 * Runtime: 8 ms
 * Memory Usage: 10.2 MB 
 */
class Solution {
public:
    int calculate(string s) {
        string formula = s + '$';
        vector<long long> nums;
        long long num = 0;
        char sign = '+';
        for (auto& c: formula) {
            if (c == ' ') continue;
            if (isdigit(c)) num = num * 10 + c - '0';
            else {
                if (sign == '+') nums.push_back(num);
                else if (sign == '-') nums.push_back(-num);
                else if (sign == '*') nums[nums.size() - 1] = nums.back() * num;
                else if (sign == '/') nums[nums.size() - 1] = nums.back() / num;
                num = 0;
                sign = c;
            }
        }
        return accumulate(nums.begin(), nums.end(), 0);
    }
};

/**
 * 109 / 109 test cases passed.
 * Runtime: 8 ms
 * Memory Usage: 9.8 MB 
 */
class Solution {
public:
    int calculate(string s) {
        vector<long> nums;
        long num = 0;
        char sign = '+';
        for (int i = 0; i < s.size(); ++ i) {
            char c = s[i];
            if (isdigit(c)) num = num * 10 + c - '0';
            if (!isdigit(c) && c != ' ' || i == (s.size() - 1)) {
                if (sign == '+') nums.push_back(num);
                else if (sign == '-') nums.push_back(-num);
                else if (sign == '*') nums.back() *= num;
                else if (sign == '/') nums.back() /= num;
                num = 0;
                sign = c;
            }
        }
        return accumulate(nums.begin(), nums.end(), 0);
    }
};

